[Sema] Remove separate closure type-checking logic

hamishknight opened the pull request:
`participatesInInference` is now always true for a non-empty body, remove it along with the separate type-checking logic such that empty bodies are type-checked together with the context.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -5515,6 +5533,55 @@ bool MissingArgumentsFailure::diagnoseClosure(const ClosureExpr *closure) {
+    return false;
+
+  auto *decl = overload->choice.getDeclOrNull();
+  if (!decl || decl->getBaseName() != ctx.Id_buildBlock)
```
There is a method we use already to determine that based on expression itself - `isResultBuilderMethodReference` all you'd need to do is check whether it's also `Id_buildBlock` specifically as well. 

For completeness you might also want to cover `Id_buildPartialBlock` I think.

hamishknight replied:
Ah nice, updated to use `isResultBuilderMethodReference`. `buildPartialBlock` is only used for non-empty cases I believe, so isn't applicable at least to the current case being handled here.

hamishknight committed:
[CS] Add custom diagnostic for missing result builder element

If we have a single missing argument for an empty
`buildBlock` call, emit a custom diagnostic.

hamishknight committed:
[Sema] Remove separate closure type-checking logic

`participatesInInference` is now always true for
a non-empty body, remove it along with the separate
type-checking logic such that empty bodies are
type-checked together with the context.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please smoke test macOS
